CollisionOperator/CollisionProcessors/neutronMGimp_class.f90

```diff
@@ -0,0 +1,371 @@
+  !!
+  subroutine capture(self, p, collDat, thisCycle, nextCycle)
+    class(neutronMGimp), intent(inout)   :: self
+    class(particle), intent(inout)       :: p
+    type(collisionData), intent(inout)   :: collDat
+    class(particleDungeon),intent(inout) :: thisCycle
+    class(particleDungeon),intent(inout) :: nextCycle
+
+    p % isDead = .true.
+
+  end subroutine capture
+
+  !!
+  !! Preform fission
+  !!
+  subroutine fission(self, p, collDat, thisCycle, nextCycle)
+    class(neutronMGimp), intent(inout)   :: self
+    class(particle), intent(inout)       :: p
+    type(collisionData), intent(inout)   :: collDat
+    class(particleDungeon),intent(inout) :: thisCycle
+    class(particleDungeon),intent(inout) :: nextCycle
+
+    p % isDead = .true.
+
+  end subroutine fission
+
+  !!
+  !! Applay cutoffs or post-collision implicit treatment
+  !!
+  subroutine cutoffs(self, p, collDat, thisCycle, nextCycle)
+    class(neutronMGimp), intent(inout)   :: self
+    class(particle), intent(inout)       :: p
+    type(collisionData), intent(inout)   :: collDat
+    class(particleDungeon),intent(inout) :: thisCycle
+    class(particleDungeon),intent(inout) :: nextCycle
+    real(defReal), dimension(3)          :: val
+    real(defReal)                        :: minWgt, maxWgt, avWgt
+
+    if (p % isDead) then
+      ! Do nothing !
+
+    ! Weight Windows treatment
+    elseif (self % weightWindows) then
+      val = self % weightWindowsMap % at(p)
+      minWgt = val(1)
+      maxWgt = val(2)
+      avWgt  = val(3)
+
+      ! If a particle is outside the WW map and all the weight limits
+      ! are zero nothing happens. NOTE: this holds for positive weights only
+      if ((p % w > maxWgt) .and. (maxWgt /= ZERO) .and. (p % splitCount < maxSplit)) then
+        call self % split(p, thisCycle, maxWgt)
+      elseif (p % w < minWgt) then
+        call self % russianRoulette(p, avWgt)
+      end if
+
+    end if
+
+  end subroutine cutoffs
+
+  !!
+  !! Perform Russian roulette on a particle
+  !!
+  subroutine russianRoulette(self, p, avWgt)
+    class(neutronMGimp), intent(inout) :: self
+    class(particle), intent(inout)     :: p
+    real(defReal), intent(in)          :: avWgt
+
+    if (p % pRNG % get() < (ONE - p % w/avWgt)) then
+      p % isDead = .true.
+    else
+      p % w = avWgt
+    end if
+
+  end subroutine russianRoulette
+
+  !!
+  !! Split particle which has too large a weight
+  !!
+  subroutine split(self, p, thisCycle, maxWgt)
+    class(neutronMGimp), intent(inout)    :: self
+    class(particle), intent(inout)        :: p
+    class(particleDungeon), intent(inout) :: thisCycle
+    real(defReal), intent(in)             :: maxWgt
+    integer(shortInt)                     :: mult, i, splitCount
+
+    ! This value must be at least 2
+    mult = ceiling(p % w/maxWgt)
+
+    ! Limit maximum split
+    if (mult > maxSplit - p % splitCount + 1) then
+      mult = maxSplit  - p % splitCount + 1
+    end if
```

Comment:
Same comments as in CE case: Perhaps the meaning of the condition in the `if` may be clearer if restructured?